src/components/network/network-loader.js

```diff
@@ -19,117 +19,255 @@ genotet.utils.inherit(genotet.NetworkLoader, genotet.ViewLoader);
+    return;
+  }
+
   switch (method) {
     case 'set':
-      // Totally replace the regex.
-      regex = geneRegex;
+      this.loadNetwork_(this.data.network.fileName, genes);
       break;
     case 'add':
-      // Concat the two regex's. We need to include the previously existing
-      // genes too so as to find the edges between the new genes and old
-      // ones.
-      regex += '|' + geneRegex;
+      this.addGenes_(genes);
       break;
     case 'remove':
-      // Remove genes do not need communication with the server.
-      this.removeGenes_(geneRegex);
-      // Return immediately. No ajax.
-      return;
+      this.deleteGenes_(genes);
+      this.signal('hideNode', {
+        genes: genes
+      });
+      break;
   }
-  this.load(this.data.fileName, regex);
 };
 
+/**
+ * Fetches the incident edges of a given node.
+ * @param {!genotet.NetworkNode} node Node of which the incident
+ * edges are queried.
+ */
+genotet.NetworkLoader.prototype.incidentEdges = function(node) {
+  var params = {
+    type: 'incident-edges',
+    fileName: this.data.network.fileName,
+    gene: node.id
+  };
+  this.get(genotet.data.serverURL, params, function(data) {
+    this.data.incidentEdges = data;
+    this.signal('incidentEdges');
+  }.bind(this), 'cannot get incident edges');
+};
 
 /**
- * Removes the selected genes from the current network data.
- * @param {string} geneRegex Regex selecting the genes to be removed.
+ * Adds genes to the network, and the edges to the original graph.
+ * @param {!Array<string>} genes Names of the genes.
  * @private
  */
-genotet.NetworkLoader.prototype.removeGenes_ = function(geneRegex) {
-  var regex;
-  try {
-    regex = RegExp(geneRegex, 'i');
-  } catch (e) {
-    genotet.error('invalid gene regex', geneRegex);
-    return;
-  }
-  this.data.nodes = _.filter(this.data.nodes, function(node) {
-    return !node.id.match(regex);
+genotet.NetworkLoader.prototype.addGenes_ = function(genes) {
+  var oldGenes = {};
+  this.data.network.nodes.forEach(function(node) {
+    oldGenes[node.id] = true;
   });
-  this.data.edges = _.filter(this.data.edges, function(edge) {
-    return !edge.source.match(regex) && !edge.target.match(regex);
+  var newGenes = [];
+  newGenes = genes.filter(function(gene) {
+    return !(gene in oldGenes);
   });
 
-  this.updateGeneRegex_();
-  this.signal('geneRemove');
+  if (!newGenes.length) {
+    genotet.warning('genes are already in the network');
+    return;
+  }
+
+  var params = {
+    type: 'incremental-edges',
+    fileName: this.data.network.fileName,
+    genes: newGenes,
+    nodes: this.data.network.nodes
+  };
+
+  this.get(genotet.data.serverURL, params, function(data) {
+    var isTF = {};
+    this.data.networkInfo.nodes.forEach(function(node) {
+      isTF[node.id] = node.isTF;
+    });
+    this.data.network.nodes = this.data.network.nodes
+      .concat(newGenes.map(function(gene) {
+      return {
+        id: gene,
+        label: gene,
```

Comment:
The id/labels in the network could be an issue.

Right now we are not taking labels from the network data. Therefore ids and labels seem to be the same thing. However there might be the case that they are different (even labels are not specifically in the data). For example, it is possible that we use lowercase gene names as their ids, but display the cased/uppercased gene names as their labels. IMO we should not assume they are the same. @jasmine